internal/impl/openai/chat_processor.go

```diff
@@ -0,0 +1,134 @@
+		panic(err)
+	}
+}
+
+func chatProcessorConfig() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Categories("AI").
+		Summary("Processor that uses the OpenAI API to create a model response for the given chat conversation.").
+		Description(`
+This processor calls the OpenAI API, creating a chat completion based on the user prompt. By default the entire message's payload as a string is submitted, and the `+"`"+ocpFieldUserPrompt+"`"+` configuration field allows customizing that.
+
+You can learn more about chat completion here: https://platform.openai.com/docs/guides/chat-completions[https://platform.openai.com/docs/guides/chat-completions^]`).
+		Version("4.31.0").
+		Fields(
+			baseConfigFieldsWithModels(
+				"gpt-4o",
+				"gpt-4o-mini",
+				"gpt-4",
+				"gpt4-turbo",
+			)...,
+		).
+		Fields(
+			service.NewBloblangField(ocpFieldUserPrompt).
+				Description("The prompt to generate text from. By default the entire payload as a string is submitted.").
+				Optional(),
+			service.NewInterpolatedStringField(ocpFieldSystemPrompt).
+				Description("The system prompt to submit along with the prompt.").
+				Optional(),
+		)
+}
+
+func makeChatProcessor(conf *service.ParsedConfig, mgr *service.Resources) (service.Processor, error) {
+	b, err := makeBaseProcessor(conf)
+	if err != nil {
+		return nil, err
+	}
+	var up *bloblang.Executor
+	if conf.Contains(ocpFieldUserPrompt) {
+		up, err = conf.FieldBloblang(ocpFieldUserPrompt)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var sp *service.InterpolatedString
+	if conf.Contains(ocpFieldSystemPrompt) {
+		sp, err = conf.FieldInterpolatedString(ocpFieldSystemPrompt)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &chatProcessor{b, up, sp}, nil
+}
+
+type chatProcessor struct {
+	*baseProcessor
+
+	userPrompt   *bloblang.Executor
+	systemPrompt *service.InterpolatedString
+}
+
+func (p *chatProcessor) Process(ctx context.Context, msg *service.Message) (service.MessageBatch, error) {
+	var body oai.ChatCompletionsOptions
+	body.DeploymentName = &p.model
+	if p.systemPrompt != nil {
+		s, err := p.systemPrompt.TryString(msg)
+		if err != nil {
+			return nil, err
+		}
+		body.Messages = append(body.Messages, &oai.ChatRequestSystemMessage{
+			Content: &s,
+		})
+	}
+	if p.userPrompt != nil {
+		s, err := msg.BloblangQueryValue(p.userPrompt)
+		if err != nil {
+			return nil, err
+		}
+		body.Messages = append(body.Messages, &oai.ChatRequestUserMessage{
+			Content: oai.NewChatRequestUserMessageContent(bloblang.ValueToString(s)),
+		})
+	} else {
+		b, err := msg.AsBytes()
+		if err != nil {
+			return nil, err
+		}
+		body.Messages = append(body.Messages, &oai.ChatRequestUserMessage{
+			Content: oai.NewChatRequestUserMessageContent(string(b)),
+		})
+	}
+	var opts oai.GetChatCompletionsOptions
+	resp, err := p.client.GetChatCompletions(ctx, body, &opts)
+	if err != nil {
+		return nil, err
+	}
+	msg = msg.Copy()
+	msg.SetBytes([]byte(*resp.Choices[0].Message.Content))
```

Comment:
We should add some checks here for the following:
- `resp` isn't nil
- `*resp.Choices` has at least one item
- `*resp.Choices[0].Message` isn't nil
- `*resp.Choices[0].Message.Content` isn't nil